napi-inl.h

```diff
@@ -1779,6 +1779,103 @@ inline Value EscapableHandleScope::Escape(Value escapee) {
   return Value(_env, result);
 }
 
+
+////////////////////////////////////////////////////////////////////////////////
+// AsyncWorker class
+////////////////////////////////////////////////////////////////////////////////
+
+inline AsyncWorker::AsyncWorker(const Function& callback) : _env(callback.Env()) {
+  _persistent = Napi::Persistent(Object::New(Env()));
+  _work = napi_create_async_work();
+}
+
+inline AsyncWorker::~AsyncWorker() {
+  if (_work != nullptr) {
+    napi_delete_async_work(_work);
+    _work = nullptr;
+  }
+}
+
+inline AsyncWorker::AsyncWorker(AsyncWorker&& other) {
+  _env = other._env;
+  other._env = nullptr;
+  _work = other._work;
+  other._work = nullptr;
+  _persistent = std::move(other._persistent);
+  _errmsg = std::move(other._errmsg);
+}
+
+inline AsyncWorker& AsyncWorker::operator =(AsyncWorker&& other) {
+  _env = other._env;
+  other._env = nullptr;
+  _work = other._work;
+  other._work = nullptr;
+  _persistent = std::move(other._persistent);
+  _errmsg = std::move(other._errmsg);
+}
+
+inline AsyncWorker::operator napi_work() const {
+  return _work;
+}
+
+inline Env AsyncWorker::Env() const {
+  return Napi::Env(_env);
+}
+
+inline void AsyncWorker::Queue() {
+  napi_async_set_data(_work, static_cast<void*>(this));
+  napi_async_set_execute(_work, OnExecute);
+  napi_async_set_complete(_work, OnWorkComplete);
+  napi_async_set_destroy(_work, OnDestroy);
+  napi_async_queue_worker(_work);
+}
+
+inline void AsyncWorker::WorkComplete() {
+  if (_errmsg.size() == 0) {
+    HandleOKCallback();
+  }
+  else {
+    HandleErrorCallback();
+  }
+}
+
+inline Object AsyncWorker::Persistent() {
+  return _persistent.Value();
+}
```

Comment:
I think I like having the Set and Get as opposed to just returning the persistentl. I'm assuming they were in NaN because they were used frequently enough that having the code once on AsyncWorker made sense.